### DISCUSSION TITLE
Swap with hibernation

### DISCUSSION THREAD

**Discussion by lucmp** [SpiralLinux/SpiralLinux-project#288] [Q&A]
After extensive tests on a virtual machine, I feel like I am ready to install Spiral on bare metal.

I see the installer offers the options No Swap, Swap with Hibernation and Swap without Hibernation.

If I enable swap with hibernation, how much space will be reserved for the swap? Does someone here know?

**Comment by geckolinux** [comment]
This is the relevant part of the Calamares configuration:

```
# Autogenerated swap sizes are as follows:
#    - *suspend*: Swap is always at least total memory size,
#      and up to 4GiB RAM follows the rule-of-thumb 2 * memory;
#      from 4GiB to 8 GiB it stays steady at 8GiB, and over 8 GiB memory
#      swap is the size of main memory.
#    - *small*: Follows the rules above, but Swap is at
#      most 8GiB, and no more than 10% of available disk.
# In both cases, a fudge factor (usually 10% extra) is applied so that there
# is some space for administrative overhead (e.g. 8 GiB swap will allocate
# 8.8GiB on disk in the end).
#
# If *file* is enabled here, make sure to have the *fstab* module
# as well (later in the exec phase) so that the swap file is
# actually created.
userSwapChoices:
    - none      # Create no swap, use no swap
    - small     # Up to 4GB
    - suspend   # At least main memory size
    # - reuse     # Re-use existing swap, but don't create any (unsupported right now)
    - file      # To swap file instead of partition

**Reply by lucmp** [reply]
Thanks. Just another question:
I chose to encrypt the entire disk and I can tell that swap is enabled, but I don't know where it is. I can see it's mounted, but it seems to be some kind of virtual block device.
Now, the installer made it the size of my RAM. What if I upgrade the RAM in the future, how do I readjust the swap size?

**Reply by geckolinux** [reply]
Hi there, please post the output of `lsblk` . You're probably seeing the zram swap, which is 95% of your RAM size per the SpiralLinux default.

**Reply by lucmp** [reply]
Output:

```
# lsblk
NAME                                          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                                             8:0    0   6,1G  0 disk  
└─sda1                                          8:1    0   6,1G  0 part  
  └─luks-f36c3751-bec4-4a29-a63e-22b293227e7a 254:0    0   6,1G  0 crypt /var/log
sr0                                            11:0    1  1024M  0 rom   
zram0                                         253:0    0 926,7M  0 disk  [SWAP]
```

95% of RAM doesn't sound good. Would that be enough for suspend to disk?

**Reply by geckolinux** [reply]
> 95% of RAM doesn't sound good.

It's probably not what you think. It doesn't mean it's going to reserve 95% of your RAM for swap.
- https://wiki.archlinux.org/title/Zram#Usage_as_swap
- https://opensource.com/article/22/11/zram-swap-linux

> Would that be enough for suspend to disk?

No amount of zRAM swap will work for hibernation, because the RAM gets cleared when powering off the machine.

You will need to use the "suspend" option in Calamares to have successful suspend-to-disk, which will create a swap partition at least the size of your amount of RAM.

**Reply by lucmp** [reply]
I did select the suspend option. Are you telling me I don't have a swap proper?

**Reply by lucmp** [reply]
I suspected the swap wasn't created because the total disk space was too small so I reinstalled everything. I'm still testing on a VM.

I can clearly see a swap partition now. Its size is TWICE the amount of RAM though. I think it's overkill, but... :-( 

But I still have the zRAM swap thing. What is it exactly? I am not familiar with it. Is it like tmpfs? The same with a different name?

**Reply by lucmp** [reply]
And two more questions if you don't mind:

1. Why does lsblk show sda1 mounted on /var/log? I expected that partition to be mounted as /.
2. Why is there a crypto_keyfile.bin file in my / directory? That is a key to what? I understand luks keyfiles, but I am used to making them myself.

**Reply by geckolinux** [reply]
> Its size is TWICE the amount of RAM though.

> `# Autogenerated swap sizes are as follows:`
> `#    - *suspend*: Swap is always at least total memory size,`
> `#      and up to 4GiB RAM follows the rule-of-thumb 2 * memory;`

> But I still have the zRAM swap thing. What is it exactly?

https://wiki.archlinux.org/title/Zram

**Comment by geckolinux** [comment]
> And two more questions if you don't mind:
> 
>     1. Why does lsblk show sda1 mounted on /var/log? I expected that partition to be mounted as /.
> 
>     2. Why is there a crypto_keyfile.bin file in my / directory? That is a key to what? I understand luks keyfiles, but I am used to making them myself.

Good questions. Sorry I don't use encryption and I've never looked into how it works.
